### PR TITLE
[data-grid] Retain Row and Column ID's when rows and columns are removed

### PR DESCRIPTION
**Pull request by kenk2**
### Summary

**What was changed:**
The Row ID is now saved whenever a user interacts with the data grid. 

**Why it was changed:**
To fix errors with accessing indexes that are no longer valid when filtering or deleting rows from the data grid.

### Testing

This change was tested using:

- [x] WDIO
- [ ] Jest
- [ ] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:

- [ ] UX review
- [ ] Accessibility review
- [x] Functional review

### Additional Details

**This PR resolves:**

UXPLATFORM-9815 

---

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 1]
```diff
@@ -279,6 +280,7 @@ const DataGrid = injectIntl((props) => {
     setFocusedCol(columnIndex);
 
     if (onColumnSelect) {
+      setLastActiveRowId('');
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:283]: The column header is a valid row as well. I would not expect us to have an empty value because we selected a column header.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 2]
```diff
@@ -476,7 +484,19 @@ const DataGrid = injectIntl((props) => {
     if (!event.currentTarget.contains(event.relatedTarget)) {
       // Not triggered when swapping focus between children
       if (handleFocus.current) {
-        setFocusedRowCol(focusedRow, focusedCol, true);
+        // When modifying the number of rows, we set the focused cell to the last selected RowId and columnId
+        if (lastActiveRowId) {
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:488]: You have logic for the rows, but there does not seem to be logic to ensure that the column is valid.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 3]
```diff
@@ -459,6 +463,10 @@ const DataGrid = injectIntl((props) => {
       event.preventDefault(); // prevent the page from moving with the arrow keys.
       return;
     }
+
+    const newRowId = nextRow === 0 ? undefined : dataGridRows[nextRow - 1]?.id;
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:467]: I don't think this logic will work with sections.  You might have to get the row identifier from the table row element.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 4]
```diff
@@ -459,6 +463,10 @@ const DataGrid = injectIntl((props) => {
...
+    setLastActiveRowId(newRowId);
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:468]: Shouldn't we be setting this value where we set the focused row and column?  I wouldn't expect them to be separated.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 5]
```diff
@@ -181,6 +181,7 @@ const DataGrid = injectIntl((props) => {
   const [checkResizable, setCheckResizable] = useState(false);
   const [focusedRow, setFocusedRow] = useState(0);
   const [focusedCol, setFocusedCol] = useState(0);
+  const [lastActiveRowId, setLastActiveRowId] = useState();
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:184]: I think that this could be a reference variable.

File: packages/terra-framework-docs/src/terra-dev-site/test/data-grid/worklist-data-grid/WorklistDataGridDeleteRows.test.jsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  const onRowSelectAll = useCallback(() => {
+    const newRowData = rowData.map(row => ({ ...row, isSelected: true }));
+    setRowData(newRowData);
+  }, [rowData]);
+
+  return (
+    <div>
+      <div>
+        <button type="button" id="delete-rows-button" onClick={() => setRowData(rowData.filter(row => !row.isSelected))}>Delete Rows</button>
+        <button type="button" id="filter-rows-button" onClick={() => setRowData(rowData.filter(row => row.isSelected))}>Filter Rows</button>
```
Comment (cm9361) [inline packages/terra-framework-docs/src/terra-dev-site/test/data-grid/worklist-data-grid/WorklistDataGridDeleteRows.test.jsx:33]: I am not clear on how this is a "Filter" button.  It seems like you should only need the one delete button.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 7]
```diff
@@ -178,6 +178,9 @@ const DataGrid = injectIntl((props) => {
   const tableContainerRef = useRef();
   const handleFocus = useRef(true);
 
+  const lastActiveRowId = useRef();
+  const lastActiveColumnId = useRef();
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:182]: Since the is a reference variable, I think you can just create one single object with the two properties.  Also, I still think we should use "focused" instead of "active".  They have two different meanings in CSS.  It would also be easier to follow if the naming conventions were the same.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 8]
```diff
@@ -196,15 +199,18 @@ const DataGrid = injectIntl((props) => {
...
   // functions
 
-  const isRowSelectionCell = (columnIndex) => (
+  const isRowSelectionCell = useCallback((columnIndex) => (
     hasSelectableRows && columnIndex < displayedColumns.length && displayedColumns[columnIndex].id === WorklistDataGridUtils.ROW_SELECTION_COLUMN.id
-  );
+  ), [displayedColumns, hasSelectableRows]);
 
-  const setFocusedRowCol = (newRowIndex, newColIndex, makeActiveElement) => {
+  const setFocusedRowCol = useCallback((newRowIndex, newColIndex, makeActiveElement) => {
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:206]: Why do we need to add a useCallback hook here?

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 9]
```diff
@@ -196,15 +199,18 @@ const DataGrid = injectIntl((props) => {
   // -------------------------------------
   // functions
 
-  const isRowSelectionCell = (columnIndex) => (
+  const isRowSelectionCell = useCallback((columnIndex) => (
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:202]: I would not expect us to need a useCallback for such a simple function.

File: packages/terra-data-grid/tests/wdio/worklist-data-grid-spec.js [hunk 10]
```diff
@@ -396,6 +396,45 @@ Terra.describeViewports('WorklistDataGrid', ['medium', 'large'], () => {
     });
   });
 
+  describe('row deletion', () => {
```
Comment (cm9361) [inline packages/terra-data-grid/tests/wdio/worklist-data-grid-spec.js:399]: I would think that you could test most of these scenarios with jest.  Do we really need as many WDIO tests?

File: packages/terra-table/src/subcomponents/ColumnHeader.jsx [hunk 11]
```diff
@@ -86,6 +86,7 @@ const ColumnHeader = (props) => {
       <tr
         aria-rowindex={1}
         className={cx('column-header-row', { hidden: !hasColumnHeaders })}
+        data-row-id="table-header-row"
```
Comment (cm9361) [inline packages/terra-table/src/subcomponents/ColumnHeader.jsx:89]: Would not be a correct row id.  If you had two Tables on the page, you would have the same id and the DOM would be broken.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 12]
```diff
@@ -476,7 +482,23 @@ const DataGrid = injectIntl((props) => {
     if (!event.currentTarget.contains(event.relatedTarget)) {
       // Not triggered when swapping focus between children
       if (handleFocus.current) {
-        setFocusedRowCol(focusedRow, focusedCol, true);
+        // When modifying the number of rows, we set the focused cell to the last selected RowId and columnId
+        if (lastActiveRowId.current && lastActiveColumnId.current) {
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:486]: I would think that you would deal with the row and column separately.  I am not sure we need both to exist to operate on one.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 13]
```diff
@@ -476,7 +482,23 @@ const DataGrid = injectIntl((props) => {
...
+          let newRowIndex = dataGridRows.findIndex(row => row.id === lastActiveRowId.current);
+          let newColumnIndex = displayedColumns.findIndex(col => col.id === lastActiveColumnId.current);
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:488]: Can we spell out "column" to avoid unnecessary abbreviations?

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 14]
```diff
@@ -276,28 +282,27 @@ const DataGrid = injectIntl((props) => {
 
   const handleColumnSelect = useCallback((columnId) => {
     const columnIndex = displayedColumns.findIndex(column => column.id === columnId);
-    setFocusedCol(columnIndex);
+    setFocusedRowCol(focusedRow, columnIndex);
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:285]: For a column selection, the row would be 0, right?

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 15]
```diff
@@ -476,7 +482,23 @@ const DataGrid = injectIntl((props) => {
     if (!event.currentTarget.contains(event.relatedTarget)) {
       // Not triggered when swapping focus between children
       if (handleFocus.current) {
-        setFocusedRowCol(focusedRow, focusedCol, true);
+        // When modifying the number of rows, we set the focused cell to the last selected RowId and columnId
+        if (lastActiveRowId.current && lastActiveColumnId.current) {
+          let newRowIndex = dataGridRows.findIndex(row => row.id === lastActiveRowId.current);
+          let newColumnIndex = displayedColumns.findIndex(col => col.id === lastActiveColumnId.current);
+
+          newRowIndex = newRowIndex === -1
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:490]: You should add small comments to your logical blocks of code.

**Comment by cm9361** [comment]
How are you testing your column boundary conditions?

File: packages/terra-data-grid/tests/wdio/worklist-data-grid-spec.js [hunk 16]
```diff
@@ -396,6 +396,54 @@ Terra.describeViewports('WorklistDataGrid', ['medium', 'large'], () => {
...
+
+    it('retains the last selected row index when filtering rows', () => {
+      clickCell(2, 1, deleteRowsSelector);
+      clickCell(8, 1, deleteRowsSelector);
+      clickCell(5, 1, deleteRowsSelector);
+
+      browser.$('#filter-rows-button').click();
+      browser.keys(['Tab']);
+
+      expect(browser.$('tbody tr:nth-child(2) th').isFocused()).toBe(true);
```
Comment (nikhitasharma) [inline packages/terra-data-grid/tests/wdio/worklist-data-grid-spec.js:413]: @kenk2  Are we intentionally not doing the snapshot matches for these tests?

Comment (nikhitasharma) [reply packages/terra-data-grid/tests/wdio/worklist-data-grid-spec.js:413]: That is fine, since visual validation may not be important in this case.

**Review by nikhitasharma — APPROVED** [review]

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 17]
```diff
@@ -195,6 +195,8 @@ const DataGrid = forwardRef((props, ref) => {
   const tableContainerRef = useRef();
   const handleFocus = useRef(true);
 
+  const lastFocusedIds = useRef({ rowId: '', columnId: '' });
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:198]: I would just name this variable, focusedCell.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 18]
```diff
@@ -495,7 +502,27 @@ const DataGrid = forwardRef((props, ref) => {
     if (!event.currentTarget.contains(event.relatedTarget)) {
       // Not triggered when swapping focus between children
       if (handleFocus.current) {
-        setFocusedRowCol(focusedRow, focusedCol, true);
+        let newRowIndex = focusedRow;
+        let newColumnIndex = focusedCol;
+        const headerOffset = hasVisibleColumnHeaders ? 1 : 0;
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:507]: I would not expect us to need a headerOffset.  Even though the headers are not visible, they are still part of the table.  Do you have a unit test for this scenario?

File: packages/terra-table/src/subcomponents/ColumnHeader.jsx [hunk 19]
```diff
@@ -84,6 +84,8 @@ const ColumnHeader = (props) => {
   return (
     <thead>
       <tr
+        aria-rowindex={1}
+        data-row-id={`${tableId}-header-row`}
```
Comment (cm9361) [inline packages/terra-table/src/subcomponents/ColumnHeader.jsx:88]: There should be Jest updates based on this change.  You should be testing that these value exist, as expected.

File: packages/terra-framework-docs/src/terra-dev-site/test/data-grid/worklist-data-grid/WorklistDataGridDeleteRowsAndColumns.test.jsx [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+      setColumnData(newColumns);
+      setRowData(newRows);
+    }
+  };
+
+  return (
+    <div>
+      <div>
+        <button type="button" id="delete-rows-button" onClick={() => setRowData(rowData.filter(row => !row.isSelected))}>Delete Rows</button>
+        <button type="button" id="filter-rows-button" onClick={() => setRowData(rowData.filter(row => row.isSelected))}>Filter Rows</button>
```
Comment (cm9361) [inline packages/terra-framework-docs/src/terra-dev-site/test/data-grid/worklist-data-grid/WorklistDataGridDeleteRowsAndColumns.test.jsx:54]: How are you testing removing the column with the selection?

**Comment by chrismichalewicz** [comment]
@kenk2  This has been functionally reviewed.

File: packages/terra-data-grid/CHANGELOG.md [hunk 21]
```diff
@@ -5,6 +5,7 @@
 * Fixed
   * Fixed issue where focus was given to the column header instead of its button element.
   * Fixed issue where row selection was being announced twice in Worklist Data Grid.
+  * Fixed an issue when data accesses invalid indexes when deleting rows and columns.
```
Comment (cm9361) [inline packages/terra-data-grid/CHANGELOG.md:8]: The issue is not specific to rows or columns being "deleted".  The data for the table is being updated and the row and/or column index no longer exists.

File: packages/terra-table/src/subcomponents/ColumnHeader.jsx (hunk 19 above)
Comment (cm9361) [reply packages/terra-table/src/subcomponents/ColumnHeader.jsx:88]: It appears you missed validation for the aria-rowindex.

File: packages/terra-table/tests/jest/ColumnHeader.test.jsx [hunk 22]
```diff
@@ -178,8 +182,11 @@ describe('ColumnHeader', () => {
     );
 
     // Verify that column headers are present
-    const columnHeader = wrapper.find('.hidden');
-    expect(columnHeader).toHaveLength(0);
+    const hiddenSelector = wrapper.find('.hidden');
+    expect(hiddenSelector).toHaveLength(0);
+
+    const columnHeader = wrapper.find('.column-header-row');
+    expect(columnHeader.props()['data-row-id']).toBe('test-table-header-row');
```
Comment (cm9361) [inline packages/terra-table/tests/jest/ColumnHeader.test.jsx:189]: You probably don't need to add the check to this test.  The purpose of this test is just to ensure the column headers are not hidden.

**Review by cm9361 — APPROVED** [review]
